File: tools/bash.py
import os
import re
import subprocess
import sys

LINUX_SUBSTRINGS = [
    "bash -c", "/bin/bash", "sh -c", "/bin/sh",
    "/usr/", "/etc/", "/home/", "/var/",
]

LINUX_COMMANDS = {
    "which", "uname", "chmod", "chown", "apt", "apt-get", "yum", "dnf",
    "head", "tail", "grep", "sed", "awk", "xargs", "touch", "rm", "mv", "cp",
}

PS_HELP = (
    "This is Windows. Use PowerShell syntax. "
    "Common mappings: `ls` = `Get-ChildItem`, `cat` = `Get-Content`, "
    "`which` = `Get-Command`, `head` = `Select-Object -First`, "
    "`grep` = `Select-String`, `uname` = `$PSVersionTable`, "
    "`python3` = `python`, `pip3` = `pip`."
)

MAX_STDOUT = 10000
MAX_STDERR = 5000


def run_bash(command: str, timeout: int = 30, cwd: str = None) -> str:
    """Run a shell command on the user's machine and return the output.
    Use this for: git operations, running builds, executing tests, file operations.
    Be careful: this can modify the system.

    Args:
        command: The shell command to execute (PowerShell syntax on Windows)
        timeout: Maximum execution time in seconds (default 30, max 120)
        cwd: Directory to run in. Defaults to the process working directory.
    """
    try:
        timeout = int(timeout)
    except (TypeError, ValueError):
        timeout = 30
    # Bound outside the try block: referencing it from the TimeoutExpired
    # handler raised UnboundLocalError when the failure happened earlier.
    safe_timeout = max(1, min(timeout, 120))

    if command is None or not str(command).strip():
        return "Error: empty command"
    command = str(command)

    # An explicit cwd is passed by the agent instead of mutating the process
    # working directory, which is global state shared with the TUI thread.
    work_dir = cwd or os.getcwd()
    if not os.path.isdir(work_dir):
        return f"Error: working directory does not exist: {work_dir}"

    try:
        if sys.platform == "win32":
            command = _sanitize_cmd(command)

            hint = _check_linux_hints(command)
            if hint:
                return f"Error: {hint}\n{PS_HELP}"

            # PowerShell is the documented shell for this agent, and cmd.exe
            # mangles nested quotes: `python -c "import x"` reaches Python as
            # a broken string literal. Route everything through PowerShell.
            shell_cmd = ["powershell", "-NoProfile", "-NonInteractive", "-Command", command]
        else:
            shell_cmd = ["sh", "-c", command]

        result = subprocess.run(
            shell_cmd,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=safe_timeout,
            cwd=work_dir,
            # A command that waits on stdin would otherwise hang until the
            # timeout with no indication of why.
            stdin=subprocess.DEVNULL,
        )

        output_parts = []
        if result.stdout:
            stdout = result.stdout.rstrip()
            if len(stdout) > MAX_STDOUT:
                stdout = stdout[:MAX_STDOUT] + "\n(truncated)"
            output_parts.append(stdout)
        if result.stderr:
            err = result.stderr.rstrip()
            if result.returncode != 0:
                lines = err.splitlines()
                if len(lines) > 15:
                    err = "(truncated)\n" + "\n".join(lines[-15:])
                elif len(err) > 2500:
                    err = "(truncated)\n" + err[-2500:].lstrip()
            output_parts.append(f"--- stderr ---\n{err[:MAX_STDERR]}")

        if not output_parts:
            if re.search(r"\bpython\b.*\.py\b", command):
                # Running a script that prints nothing is nearly always a bug,
                # but a bare "(no output)" reads like success to a small model.
                output_parts.append(
                    "(no output — the script printed nothing. If it was meant "
                    "to display something, this is a logic bug, not a success.)"
                )
            else:
                output_parts.append("(no output)")

        output = "\n".join(output_parts)

        if result.returncode != 0:
            output = f"Exit code: {result.returncode}\n{output}"
            if "Traceback" in (result.stderr or ""):
                output += "\n\nHINT: The script crashed. Read the error, fix the bug in the .py file, then re-run."

        return output
    except subprocess.TimeoutExpired:
        return f"Error: Command timed out after {safe_timeout} seconds"
    except FileNotFoundError as e:
        return f"Error: Command not found: {e}"
    except Exception as e:
        return f"Error executing command: {e}"


def _check_linux_hints(command: str) -> str | None:
    """Reject genuine Linux/bash commands so the small model learns PowerShell.

    Matches only on the *command position* of each segment, so legitimate
    arguments like `python app.py --find x` or `git grep foo` are not rejected.
    """
    lowered = command.strip().lower()

    for frag in LINUX_SUBSTRINGS:
        if frag in lowered:
            return f"Linux command detected: '{frag.strip()}'"

    for segment in re.split(r"\|\||&&|[|;&]", lowered):
        tokens = segment.strip().split()
        if not tokens:
            continue
        head = tokens[0].strip("\"'")
        if head in LINUX_COMMANDS:
            return f"Linux command detected: '{head}'"
    return None


def _sanitize_cmd(command: str) -> str:
    """Fix harmless Windows/Linux mechanics that small models get wrong.

    Only mechanical rewrites live here (python3 -> python, /dev/null, stray cd).
    Real bash usage is rejected by _check_linux_hints instead, so the model
    gets a corrective hint rather than a silent rewrite.
    """
    cmd = command.strip()
    # Only a no-op `cd` into the workspace is dropped. Stripping every leading
    # `cd X && ...` silently changed the working directory of a legitimate
    # command ("cd build && cmake .." ran in the wrong place and failed).
    cmd = re.sub(
        r"^\s*cd\s+(?:\.|[\"']\.[\"'])\s*(?:&&|;)\s*", "", cmd
    )
    cmd = _sub_outside_quotes(r"\bpython3(?:\.\d+)?\b", "python", cmd)
    cmd = _sub_outside_quotes(r"\bpip3\b", "pip", cmd)
    cmd = re.sub(r"\s*2>\s*/dev/null", "", cmd)
    cmd = re.sub(r"\s*1?>\s*/dev/null", "", cmd)
    cmd = re.sub(r"\s+2>&1\s*\|\s*cat\b", "", cmd)
    cmd = _squeeze_spaces_outside_quotes(cmd)
    return cmd.strip().strip(";").strip()


def _sub_outside_quotes(pattern: str, repl: str, cmd: str) -> str:
    """Apply a substitution only to the unquoted parts of a command.

    Rewriting inside quotes corrupts the user's own data: a commit message or
    a here-string that legitimately mentions "python3" must survive intact.
    """
    out = []
    buf = []
    quote = None
    for ch in cmd:
        if quote:
            out.append(ch)
            if ch == quote:
                quote = None
            continue
        if ch in "\"'":
            out.append(re.sub(pattern, repl, "".join(buf)))
            buf = []
            quote = ch
            out.append(ch)
            continue
        buf.append(ch)
    out.append(re.sub(pattern, repl, "".join(buf)))
    return "".join(out)


def _squeeze_spaces_outside_quotes(cmd: str) -> str:
    """Collapse runs of whitespace, but never inside a quoted string.

    A blanket re.sub rewrote the user's own data: the commit message in
    `git commit -m "fix:  spacing"` came out with the spacing altered.
    """
    out = []
    quote = None
    prev_space = False
    for ch in cmd:
        if quote:
            out.append(ch)
            if ch == quote:
                quote = None
            continue
        if ch in "\"'":
            quote = ch
            out.append(ch)
            prev_space = False
            continue
        if ch in " \t":
            if not prev_space:
                out.append(" ")
            prev_space = True
            continue
        out.append(ch)
        prev_space = False
    return "".join(out)

File: internal/tools/bash.go
package tools

import (
	"context"
	"fmt"
	"os/exec"
	"regexp"
	"runtime"
	"strings"
	"time"
)

var linuxSubstrings = []string{
	"bash -c", "/bin/bash", "sh -c", "/bin/sh",
	"/usr/", "/etc/", "/home/", "/var/",
}

var linuxCommands = map[string]bool{
	"which": true, "uname": true, "chmod": true, "chown": true,
	"apt": true, "apt-get": true, "yum": true, "dnf": true,
	"head": true, "tail": true, "grep": true, "sed": true,
	"awk": true, "xargs": true, "touch": true, "rm": true,
	"mv": true, "cp": true,
}

const psHelp = "This is Windows. Use PowerShell syntax. " +
	"Common mappings: `ls` = `Get-ChildItem`, `cat` = `Get-Content`, " +
	"`which` = `Get-Command`, `head` = `Select-Object -First`, " +
	"`grep` = `Select-String`, `uname` = `$PSVersionTable`, " +
	"`python3` = `python`, `pip3` = `pip`."

// RunBash runs a shell command and returns its output.
func RunBash(workspace, command string, timeout int) string {
	if timeout < 1 {
		timeout = 30
	}
	if timeout > 120 {
		timeout = 120
	}

	if runtime.GOOS == "windows" {
		command = SanitizeCmd(command)
		if hint := checkLinuxHints(command); hint != "" {
			return fmt.Sprintf("Error: %s\n%s", hint, psHelp)
		}
	}

	var name string
	var args []string
	if runtime.GOOS == "windows" {
		name, args = "powershell", []string{"-NoProfile", "-Command", command}
	} else {
		name, args = "sh", []string{"-c", command}
	}

	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, name, args...)
	cmd.Dir = workspace
	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()

	if ctx.Err() == context.DeadlineExceeded {
		return fmt.Sprintf("Error: Command timed out after %d seconds", timeout)
	}

	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			return fmt.Sprintf("Error: Command not found: %v", err)
		}
	}

	var parts []string
	if s := stdout.String(); s != "" {
		s = strings.TrimRight(s, "\n")
		if len(s) > 10000 {
			s = s[:10000]
		}
		parts = append(parts, s)
	}
	if s := stderr.String(); s != "" {
		s = strings.TrimRight(s, "\n")
		if exitCode != 0 {
			lines := strings.Split(s, "\n")
			if len(lines) > 15 {
				s = "(truncated)\n" + strings.Join(lines[len(lines)-15:], "\n")
			} else if len(s) > 2500 {
				s = strings.TrimLeft(s[len(s)-2500:], " \t\n")
				s = "(truncated)\n" + s
			}
		}
		if len(s) > 5000 {
			s = s[:5000]
		}
		parts = append(parts, "--- stderr ---\n"+s)
	}

	if len(parts) == 0 {
		if pythonScriptRe.MatchString(command) {
			parts = append(parts, "(no output — the script printed nothing. If it was meant to display something, this is a logic bug, not a success.)")
		} else {
			parts = append(parts, "(no output)")
		}
	}

	output := strings.Join(parts, "\n")
	if exitCode != 0 {
		output = fmt.Sprintf("Exit code: %d\n%s", exitCode, output)
		if strings.Contains(stderr.String(), "Traceback") {
			output += "\n\nHINT: The script crashed. Read the error, fix the bug in the .py file, then re-run."
		}
	}
	return output
}

var pythonScriptRe = regexp.MustCompile(`\bpython\b.*\.py\b`)

// CheckLinuxHints reports a rejected Linux command for the model to learn from.
func CheckLinuxHints(command string) string { return checkLinuxHints(command) }

func checkLinuxHints(command string) string {
	lowered := strings.ToLower(strings.TrimSpace(command))
	for _, frag := range linuxSubstrings {
		if strings.Contains(lowered, frag) {
			return fmt.Sprintf("Linux command detected: '%s'", strings.TrimSpace(frag))
		}
	}
	for _, segment := range linuxSegmentRe.Split(lowered, -1) {
		tokens := strings.Fields(segment)
		if len(tokens) == 0 {
			continue
		}
		head := strings.Trim(tokens[0], "\"'")
		if linuxCommands[head] {
			return fmt.Sprintf("Linux command detected: '%s'", head)
		}
	}
	return ""
}

var linuxSegmentRe = regexp.MustCompile(`\|\||&&|[|;&]`)

// SanitizeCmd fixes harmless Windows/Linux mechanics small models get wrong.
func SanitizeCmd(command string) string {
	cmd := strings.TrimSpace(command)
	cmd = reLeadingCd.ReplaceAllString(cmd, "")
	cmd = rePython3.ReplaceAllString(cmd, "python")
	cmd = rePip3.ReplaceAllString(cmd, "pip")
	cmd = reDevNull2.ReplaceAllString(cmd, "")
	cmd = reDevNull.ReplaceAllString(cmd, "")
	cmd = reCatPipe.ReplaceAllString(cmd, "")
	cmd = squeezeSpacesOutsideQuotes(cmd)
	return strings.Trim(strings.TrimSpace(cmd), ";")
}

var (
	reLeadingCd = regexp.MustCompile(`^\s*cd\s+(?:\.|["']\.["'])\s*(?:&&|;)\s*`)
	rePython3   = regexp.MustCompile(`\bpython3(?:\.\d+)?\b`)
	rePip3      = regexp.MustCompile(`\bpip3\b`)
	reDevNull2  = regexp.MustCompile(`\s*2>\s*/dev/null`)
	reDevNull   = regexp.MustCompile(`\s*1?>\s*/dev/null`)
	reCatPipe   = regexp.MustCompile(`\s+2>&1\s*\|\s*cat\b`)
)

func squeezeSpacesOutsideQuotes(cmd string) string {
	var out strings.Builder
	var quote rune
	prevSpace := false
	for _, ch := range cmd {
		if quote != 0 {
			out.WriteRune(ch)
			if ch == quote {
				quote = 0
			}
			continue
		}
		if ch == '"' || ch == '\'' {
			quote = ch
			out.WriteRune(ch)
			prevSpace = false
			continue
		}
		if ch == ' ' || ch == '\t' {
			if !prevSpace {
				out.WriteRune(' ')
			}
			prevSpace = true
			continue
		}
		out.WriteRune(ch)
		prevSpace = false
	}
	return out.String()
}

type cmdToken struct {
	raw    string
	quoted bool
}

// TokenizeCmd splits a command into tokens, keeping quoted spans whole.
func TokenizeCmd(cmd string) []cmdToken {
	var tokens []cmdToken
	var cur strings.Builder
	curQuoted := false
	var inQ rune
	for _, ch := range cmd {
		if inQ != 0 {
			cur.WriteRune(ch)
			if ch == inQ {
				inQ = 0
			}
			continue
		}
		if ch == '"' || ch == '\'' {
			if cur.Len() == 0 {
				curQuoted = true
			}
			inQ = ch
			cur.WriteRune(ch)
			continue
		}
		if ch == ' ' || ch == '\t' {
			if cur.Len() != 0 {
				tokens = append(tokens, cmdToken{cur.String(), curQuoted})
				cur.Reset()
				curQuoted = false
			}
			continue
		}
		cur.WriteRune(ch)
	}
	if cur.Len() != 0 {
		tokens = append(tokens, cmdToken{cur.String(), curQuoted})
	}
	return tokens
}

// QuoteExistingPaths wraps unquoted runs of tokens that name a real workspace
// file in double quotes, so spaced paths survive the shell.
func QuoteExistingPaths(workspace, command string) string {
	tokens := TokenizeCmd(command)
	var out []string
	i, n := 0, len(tokens)
	for i < n {
		raw, wasQuoted := tokens[i].raw, tokens[i].quoted
		if wasQuoted {
			out = append(out, raw)
			i++
			continue
		}
		best := -1
		for j := i + 1; j <= n; j++ {
			window := tokens[i:j]
			anyQuoted := false
			for _, t := range window {
				if t.quoted {
					anyQuoted = true
					break
				}
			}
			if anyQuoted {
				break
			}
			joined := strings.TrimSpace(strings.Join(raws(window), " "))
			cand := joined
			if !isAbs(cand) {
				cand = joinPath(workspace, joined)
			}
			if pathExists(cand) {
				best = j
			}
		}
		if best != -1 {
			joined := strings.TrimSpace(strings.Join(raws(tokens[i:best]), " "))
			out = append(out, `"`+joined+`"`)
			i = best
		} else {
			out = append(out, raw)
			i++
		}
	}
	return strings.Join(out, " ")
}

func raws(tokens []cmdToken) []string {
	s := make([]string, len(tokens))
	for i, t := range tokens {
		s[i] = t.raw
	}
	return s
}
